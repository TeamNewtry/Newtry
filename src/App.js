/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import React from 'react';
import {View, StyleSheet, Text, TextInput} from 'react-native';

import {Colors} from 'react-native/Libraries/NewAppScreen';

import firestore from '@react-native-firebase/firestore';

const Section = ({children, title}): Node => {
  return (
    <View style={styles.sectionContainer}>
      <Text style={[styles.sectionTitle, {color: Colors.white}]}>{title}</Text>
      <Text style={[styles.sectionDescription, {color: Colors.light}]}>
        {children}
      </Text>
    </View>
  );
};

const App: () => Node = () => {
  const [name, setName] = React.useState('Retrieving...');
  const [age, setAge] = React.useState('0');

  let doc1 = firestore().collection('test').doc('1');

  // init live update of name & age
  let subscriber = doc1.onSnapshot(doc => {
    let data = doc.data();
    setName(data.name);
    setAge(data.age);
  });

  const onChangeName = name => onChangeInfo(name, false);
  const onChangeAge = name => onChangeInfo(name, true);
  const onChangeInfo = (info, isAge) => {
    doc1.update({name: isAge ? name : info, age: isAge ? info : age}).done();
  };

  return (
    <View style={{backgroundColor: Colors.black}}>
      <Section title="Doc 'test'">
        Name: <Text style={styles.highlight}>{name}</Text> {'\n'}
        Age: <Text style={styles.highlight}>{age}</Text>
      </Section>
      <Section title="Tell me more about you...">
        ...in the following form
      </Section>
      <TextInput
        style={styles.input}
        onChangeText={onChangeName}
        value={name}
        placeholder="Margarete"
      />
      <TextInput
        style={styles.input}
        onChangeText={onChangeAge}
        value={age.toString()}
        placeholder="44 oder doch 13Â½"
        keyboardType="numeric"
      />
    </View>
  );
};

const styles = StyleSheet.create({
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
  },
  highlight: {
    fontWeight: '700',
  },
  input: {
    height: 40,
    margin: 12,
    borderWidth: 1,
    borderColor: Colors.light,
    padding: 10,
  },
});

export default App;
