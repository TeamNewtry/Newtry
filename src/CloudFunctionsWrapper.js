import firebase from '@react-native-firebase/app';
import '@react-native-firebase/functions';
import Product from './store/Product';
import Nutrition from './store/Nutrition';

const functions = firebase.app().functions('europe-west1');

function getProductByGTIN(gtin) {
  return callCloudFunction('getProduct', {id: gtin}, response => {
    const info = response.data.product;
    const nutriInfo = info.nutriments;
    let nutrition = new Nutrition(
      nutriInfo['energy-kcal'],
      nutriInfo.fat,
      nutriInfo['saturated-fat'],
      nutriInfo.carbohydrates,
      nutriInfo.sugars,
      nutriInfo.proteins,
      nutriInfo.salt,
    );
    return new Product(
      info.code,
      info.product_name,
      info.generic_name,
      nutrition,
      info.ingredients_text,
      null,
      info.nutriscore_data.grade,
      [info.image_url],
    );
  });
}

function searchProducts(gtinOrName) {
  return callCloudFunction('search', {searchTerm: gtinOrName}, response => {
    const info = response.data;
    console.log(info);
    const productIds = info.map(product => product.id);
    console.log(productIds);
    return productIds;
  });
}

async function callCloudFunction(functionName, parameters, resolve) {
  return await functions
    .httpsCallable(functionName)(parameters)
    .then(response => {
      if (
        typeof response.data !== 'object' ||
        'httpErrorCode' in response.data
      ) {
        console.error(`fail ${functionName} with`, parameters);
        console.error(response.data);
        return;
      }
      return resolve(response);
    })
    .catch(reason => {
      console.error(`error ${functionName} with`, parameters);
      console.error(
        'Code: ',
        reason.code,
        'Message: ',
        reason.message,
        'Details: ',
        reason.details,
      );
    });
}

export {getProductByGTIN, searchProducts};
