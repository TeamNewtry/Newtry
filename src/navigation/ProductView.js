import React, {useContext} from 'react';
import {Image, ScrollView, StyleSheet, Text, View} from 'react-native';
import {Row, Rows, Table, TableWrapper} from 'react-native-table-component';
import Unorderedlist from 'react-native-unordered-list';
import {useAsync} from 'react-async';
import {getProductByGTIN} from '../CloudFunctionsWrapper';
import {Icon} from 'react-native-elements';
import {LocalizationContext} from '../components/Translations';
import LinearGradient from 'react-native-linear-gradient';
import NutritionTable from '../components/NutritionTable';
import IngredientList from '../components/IngredientList';
import DetailPage from '../components/DetailPage';
import CustomerRating from '../components/CustomerRating';
import Comments from '../components/comments/Comments';

const ProductView = ({navigation, route}) => {
  const {translations} = useContext(LocalizationContext);
  const {data, error} = useAsync(getProductByGTIN, {
    gtin: route.params.gtin,
  });
  console.log(JSON.stringify(data));
  if (typeof data !== 'undefined') {
    const page = (
      <ScrollView>
        <View style={styles.imageBorder}>
          <Image
            source={{uri: data.pictures.toString()}}
            style={styles.image}
          />
          <LinearGradient
            start={{x: 1, y: 0}}
            end={{x: 0, y: 0}}
            colors={['#24FF00', '#00D8D4', '#60dbfd']}
            style={styles.line}
          />
        </View>
        <View style={styles.padding}>
          <View style={styles.textContainer}>
            <Text style={styles.title}>{data.name}</Text>
            <Text style={styles.gtin}>GTIN: {data.gtin}</Text>
            <Text />
            <Text style={styles.description}>{data.description}</Text>
          </View>
          <NutritionTable data={data.nutrition} />
          <IngredientList data={data.ingredients} />
        </View>
        <LinearGradient
          start={{x: 1, y: 0}}
          end={{x: 0, y: 0}}
          colors={['#24FF00', '#00D8D4', '#60dbfd']}
          style={styles.line}
        />

        <View style={styles.padding}>
          <Text style={styles.subTitle}>
            {translations['customerrating.header']}
          </Text>
          <CustomerRating productId={data.gtin} />
        </View>
        <LinearGradient
          start={{x: 1, y: 0}}
          end={{x: 0, y: 0}}
          colors={['#24FF00', '#00D8D4', '#60dbfd']}
          style={styles.line}
        />

        <View style={styles.padding}>
          <Text style={styles.subTitle}>{translations['rating.header']}</Text>
          <Comments productId={data.gtin} />
        </View>
      </ScrollView>
    );
    return DetailPage({navigation, route}, page);
  } else {
    return <View />;
  }
};

export default ProductView;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#EAFFFA',
  },
  wrapper: {
    flexDirection: 'row',
  },
  title: {
    fontSize: 20,
    paddingTop: 10,
    fontFamily: 'Comfortaa',
  },
  subTitle: {
    fontSize: 16,
    paddingBottom: 20,
    fontFamily: 'Comfortaa',
  },
  textContainer: {marginTop: 10, marginBottom: 10},
  image: {
    width: window.outerWidth,
    height: 400,
    resizeMode: 'contain',
    backgroundColor: 'rgb(246, 246, 247)',
  },
  description: {
    fontSize: 14,
    paddingBottom: 1,
    fontFamily: 'Comfortaa',
  },
  gtin: {
    fontSize: 10,
    fontFamily: 'Comfortaa',
  },
  padding: {
    padding: 20,
  },
  line: {
    height: 2,
  },
  backArrow: {
    position: 'absolute',
    zIndex: 10,
    top: 5,
    left: 5,
  },
});
