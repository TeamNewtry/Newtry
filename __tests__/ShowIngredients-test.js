import React from 'react';
import renderer from 'react-test-renderer';
import ProductView from '../src/navigation/ProductView';
import * as MockCloudFunctionsWrapper from '../src/CloudFunctionsWrapper';

jest.mock(
  '../node_modules/react-native/Libraries/EventEmitter/NativeEventEmitter',
);
jest.mock(
  'react-native-table-component',
  () => 'Row, Rows, Table, TableWrapper',
);
jest.mock('../src/CloudFunctionsWrapper');
MockCloudFunctionsWrapper.getProductByGTIN.mockImplementation(gtin =>
  Promise.resolve({data: {gtin: gtin}}),
);

test('renders ingredients', () => {
  const route = {
    isAccessibilityModeOn: false,
    params: {
      gtin: '0000080135876',
    },
  };
  const navigate = jest.fn();
  const tree = renderer
    .create(<ProductView route={route} navigation={{navigate}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
