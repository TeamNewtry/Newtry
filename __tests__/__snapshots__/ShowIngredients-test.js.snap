// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `<View />`;

exports[`renders ingredients 1`] = `<View />`;
